backend/data_preparation/crawler/fire_crawler.py

```diff
@@ -0,0 +1,318 @@
+
+    def is_valid(self) -> bool:
+        """
+        Judges if this fire event is valid. A valid fire event should have reasonable fire name.
+        :return: bool
+        """
+        # to delete exceptions: "[To Parent Directory]" "whatever.zip"
+        # ActivePerim is an useless link which contains metadata over the year, existing in links before 2016 Skip it
+        return "." in self.url_name or "[To Parent Directory]" == self.url_name or self.url_name == "ActivePerim"
+
+    def __str__(self) -> str:
+        """
+        Transforms a FireEvent into a string.
+        :return: str
+        """
+        if id != -1:
+            return f"Fire Event {self.fire_id}: {self.url_name} in year {self.year}, state {self.state}"
+        return f"Fire Event Unknown: {self.url_name} in year {self.year}, state {self.state}"
+
+    def to_url(self) -> str:
+        """
+        Converts a FireEvent object into the url
+        :return: url. e.g. "https://rmgsc.cr.usgs.gov/outgoing/GeoMAC/2017_fire_data/California/Eclipse"
+        """
+        current_year = datetime.datetime.now().date().year
+        year_str = "current_year" if self.year == current_year else str(self.year)
+        return f"{FireCrawler.BASE_DIR}{year_str}_fire_data/{self.state}/{self.url_name}"
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Converts a FireEvent object into a dictionary.
+        :return: Dict[str, Any]
+            e.g. {"year": 2017,
+                "firename": "Sand",
+                "state": "California",
+                "id": 299,
+                "url":"https://rmgsc.cr.usgs.gov/outgoing/GeoMAC/...."}
+        """
+        data = {"year": self.year,
+                "firename": self.url_name,
+                "state": self.state,
+                "id": self.fire_id,
+                "url": self.to_url()}
+        return data
+
+    def to_tuple(self, new_id) -> Tuple[int,int,str, str]:
+        self.fire_id = new_id if self.fire_id == -1 else self.fire_id
+        return self.fire_id, self.year, self.state, self.url_name
+
+
+class FireCrawler(CrawlerBase):
+    # BASE_DIR is the website directory the crawler will crawl
+    BASE_DIR = 'https://rmgsc.cr.usgs.gov/outgoing/GeoMAC/'
+    RE_EXTRACT_FIRE_EVENTS_IN_STATE = re.compile(r'<A .*?>(.*?)</A>', re.S | re.M)
+    RE_EXTRACT_ALL_FIRES = re.compile(r'<A .*?>(\w*?)</A>', re.S | re.M)
+    RE_EXTRACT_FILES_IN_FIRE_EVENTS = re.compile(r'<A .*?>([a-z]{2}_.+?)</A>', re.S | re.M)
+
+    def __init__(self, states: List[str]):
+        """
+        Initialization for the crawler.
+        :param states: list of states that the crawler needs to crawl
+                e.g. ["California", "Nevada"]
+        """
+        super().__init__()
+        self.states = states
+        # states is a list of states that the crawler needs to crawl
+
+    @staticmethod
+    @retry(stop_max_attempt_number=3)
+    def _get_url(url: str, page_name: str, function_name: str) -> str:
```

Comment:
from design perspective, passing the function name into other functions in the scope is not a good idea. I would like you to remove this parameter. in logger setup, it will log whichever function that is calling this `_get_url` helper function. so no need to pass in function_name manually.